src/workerd/util/state-machine.h

```diff
@@ -0,0 +1,3897 @@
+  //
+  // UNSAFE PATTERN:
+  //   Active& a = machine.get<Active>();
+  //   someFunction();  // If this transitions the machine...
+  //   a.doSomething(); // ...this is use-after-free!
+  //
+  // SAFE PATTERN:
+  //   machine.withState<Active>([](Active& a) {
+  //     a.doSomething();  // Transitions blocked during callback
+  //   });
+
+  // Get reference to current state data.
+  // Throws if not in the specified state or if uninitialized.
+  //
+  // WARNING: The returned reference is NOT protected by a transition lock.
+  // It becomes dangling if the machine transitions before you're done using it.
+  // Prefer withState() for safe access.
+  template <typename S>
+  S& get() KJ_LIFETIMEBOUND
+    requires(_::isOneOf<S, States...>)
+  {
+    requireInitialized();
+    KJ_REQUIRE(is<S>(), "State machine is not in the expected state");
+    return state.template get<S>();
+  }
+
+  template <typename S>
+  const S& get() const KJ_LIFETIMEBOUND
+    requires(_::isOneOf<S, States...>)
+  {
+    requireInitialized();
+    KJ_REQUIRE(is<S>(), "State machine is not in the expected state");
+    return state.template get<S>();
+  }
+
+  // Try to get reference to current state data.
+  // Returns kj::none if not in the specified state.
+  //
+  // WARNING: The returned reference is NOT protected by a transition lock.
+  // It becomes dangling if the machine transitions before you're done using it.
+  // Prefer withState() for safe access.
+  template <typename S>
+  kj::Maybe<S&> tryGet() KJ_LIFETIMEBOUND
+    requires(_::isOneOf<S, States...>)
+  {
+    return state.template tryGet<S>();
+  }
+
+  template <typename S>
+  kj::Maybe<const S&> tryGet() const KJ_LIFETIMEBOUND
+    requires(_::isOneOf<S, States...>)
+  {
+    return state.template tryGet<S>();
+  }
+
+  // ---------------------------------------------------------------------------
+  // Safe State Access (RECOMMENDED)
+  // ---------------------------------------------------------------------------
+
+  // Execute a function with the current state, locking transitions.
+  // This is the SAFEST way to access state data as it prevents
+  // use-after-free by blocking transitions during the callback.
+  //
+  // Usage:
+  //   machine.withState<Active>([](Active& a) {
+  //     a.resource->doSomething();  // Safe!
+  //   });
+  //
+  // Returns the function's result wrapped in Maybe (none if not in state).
+  // For void functions, returns true if executed, false if not in state.
+  //
+  // WARNING: Do NOT store or escape the reference passed to the callback!
+  // The reference is only valid during the callback. Storing it for later
+  // use defeats the purpose of the transition lock:
+  //
+  //   Active* escaped;
+  //   machine.withState<Active>([&](Active& a) {
+  //     escaped = &a;  // BAD: escaping the reference!
+  //   });
+  //   escaped->foo();  // UAF if machine has transitioned!
+  template <typename S, typename Func>
+  auto withState(
```

Comment:
When I originally read the example code, I assumed `withState<T>()` would throw if the state was not `T`. I thought that, because of its asymmetry with `whenActive()` -- I figured `when` meant "if it is" and `with` meant "it must be".

I think it might help if they were both "when", or both "with", e.g.

```cpp
whenStateIs<T>(), whenActive()  // allows other states
withState<T>(), withActiveState()  // disallows other states

// or if you like "with" to be lenient

withState<T>(), withActiveState()  // allows other states
requireState<T>(), requireActiveState()  // disallows other states
```

This comment applies to all the other state machine implementations, too.